File: backend/app/routers/clients.py
"""Read-only clients API.

Exposes the entity ↔ person ↔ role assignments (the `entity_people_roles`
join) so the SPA can list who is attached to which entity and in what role.
"""

from typing import List, Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app import models, schemas
from app.auth import verify_token
from app.database import get_db

router = APIRouter(tags=["clients"], dependencies=[Depends(verify_token)])


@router.get("/clients", response_model=List[schemas.ClientContactRow])
def list_clients(
    db: Session = Depends(get_db),
    first_name: Optional[str] = Query(
        None,
        description="Case-insensitive substring match on people.first_name ('%q%').",
    ),
    last_name: Optional[str] = Query(
        None,
        description="Case-insensitive substring match on people.last_name ('%q%').",
    ),
    company: Optional[str] = Query(
        None,
        description="Case-insensitive substring match on entities.entity_name ('%q%').",
    ),
):
    # Mirrors the canonical query exactly:
    #   select epr.identity_people_roles, e.entity_name, p.first_name,
    #          p.last_name, pr.role_name, epr.created_at
    #   from entity_people_roles epr
    #   join entities e      on e.entity_id = epr.entities_entity_id
    #   join people p        on p.idperson  = epr.people_idperson
    #   join people_roles pr on pr.idroles  = epr.roles_idroles;
    q = (
        db.query(
            models.EntityPeopleRole.identity_people_roles,
            models.Entity.entity_name,
            models.Person.first_name,
            models.Person.last_name,
            models.PeopleRole.role_name,
            models.EntityPeopleRole.created_at,
        )
        .join(
            models.Entity,
            models.Entity.entity_id == models.EntityPeopleRole.entities_entity_id,
        )
        .join(
            models.Person,
            models.Person.idperson == models.EntityPeopleRole.people_idperson,
        )
        .join(
            models.PeopleRole,
            models.PeopleRole.idroles == models.EntityPeopleRole.roles_idroles,
        )
    )

    # Optional substring filters; each applied independently (combined with AND).
    if first_name and first_name.strip():
        q = q.filter(models.Person.first_name.ilike(f"%{first_name.strip()}%"))
    if last_name and last_name.strip():
        q = q.filter(models.Person.last_name.ilike(f"%{last_name.strip()}%"))
    if company and company.strip():
        q = q.filter(models.Entity.entity_name.ilike(f"%{company.strip()}%"))

    return q.all()

File: frontend/src/services/clients.ts
// Clients API service — talks to the read-only GET /clients endpoint, which
// exposes the entity ↔ person ↔ role join (entity_people_roles) for the
// "Add New Lead" client autocomplete.

import { api } from "./api";

// One row of the entity_people_roles join (backend schemas.ClientContactRow).
export interface ClientContactRow {
  /** epr id (identity_people_roles) — the hidden key sent back on lead create. */
  identity_people_roles: number;
  entity_name: string;
  first_name: string;
  last_name: string;
  role_name: string;
  /** All known emails for the person — surfaced as suggestions when >1. */
  emails: string[];
  /** All known phones for the person — surfaced as suggestions when >1. */
  phones: string[];
  created_at: string | null;
}

export const clientsApi = {
  /** Substring search across first name / last name / company. Each non-empty
   *  field is sent as a filter and the backend combines them with AND. */
  async search(params: {
    firstName?: string;
    lastName?: string;
    company?: string;
  }): Promise<ClientContactRow[]> {
    const qs = new URLSearchParams();
    if (params.firstName?.trim()) qs.set("first_name", params.firstName.trim());
    if (params.lastName?.trim()) qs.set("last_name", params.lastName.trim());
    if (params.company?.trim()) qs.set("company", params.company.trim());
    const query = qs.toString();
    return api.get<ClientContactRow[]>(`/clients${query ? `?${query}` : ""}`);
  },
};
